web-api/src/storage/elastic.rs

```diff
@@ -273,18 +274,29 @@ fn is_success_status(status: u16, allow_not_found: bool) -> bool {
         .unwrap_or(false)
 }
 
-#[async_trait]
-impl storage::Document for Storage {
-    async fn get_by_ids(&self, ids: &[&DocumentId]) -> Result<Vec<PersonalizedDocument>, Error> {
+impl Storage {
+    pub(crate) async fn get_by_ids_with_transaction(
+        &self,
+        ids: &[&DocumentId],
+        tx: Option<&mut Transaction<'_, Postgres>>,
+    ) -> Result<Vec<PersonalizedDocument>, Error> {
         if ids.is_empty() {
             return Ok(Vec::new());
         }
 
+        let values = if let Some(tx) = tx {
+            self.postgres
+                .documents_exist_with_transaction(ids, tx)
+                .await?
+        } else {
+            self.postgres.documents_exist(ids).await?
+        };
+
```

Comment:
:art: Can't we have a method on `self.postgres` to get a transaction and use that instead of calling two different methods?

Reply:
That would be:

```rust
let values = if let Some(tx) = tx {
            self.postgres
                .documents_exist_with_transaction(ids, tx)
                .await?
        } else {
            let mut tx = self.postgres.pool.begin().await?;
            self.postgres.documents_exist_with_transaction(ids, tx).await?;
            self.postgres.pool.commit().await?;
        };
```

But the `else` block is basically the implementation of `documents_exist` and that functions is used in other places anyway.

Or instead of one branch and calling the function in two places you could have two branches one before and one after the call where you conditionally create/commit the transaction. But that's more complex and bug prone.

The reason why `get_by_ids_with_transaction` is accepting an `Option` instead of just always accepting a transaction like the other functions is that in cases where we pass in `None` the would have to start the transactions before making queries to elastic search making it unnecessary longer.  At the same time it can create the illusion for the reader that it somehow at least somewhat guards the elastic search query for consistency. But due to how postgres and elastic search work this is not the case at all.

Reply:
I had considered adding some ad-hoc transaction abstraction but this overlaps with adding a general abstraction for transactions so anything like that should be done with or after creating such abstraction.